plat/common/memory.c

```diff
@@ -137,3 +139,217 @@ void *ukplat_memregion_alloc(__sz size, int type)
+ * contains the right memory region of higher priority, then split the left one
+ * in two, by adjusting the current left one and inserting a new memory region
+ * descriptor.
+ */
+static inline void overlapping_mrd_fixup(struct ukplat_memregion_list *list,
+					 struct ukplat_memregion_desc *const ml,
+					 struct ukplat_memregion_desc *const mr,
+					 int ml_prio, int mr_prio,
+					 __u32 lidx __unused, __u32 ridx)
+{
+	/* If left memory region is of higher priority */
+	if (ml_prio > mr_prio) {
+		/* If the right region is contained within the left region,
+		 * drop it entirely
+		 */
+		if (RANGE_CONTAIN(ml->pbase, ml->len, mr->pbase, mr->len)) {
+			mr->len = 0;
+
+		/* If the right region has a part of itself in the left region,
+		 * drop that part of the right region only
+		 */
+		} else {
+			mr->len -= ml->pbase + ml->len - mr->pbase;
+			mr->pbase = ml->pbase + ml->len;
+			mr->vbase = mr->pbase;
+		}
+
+	/* If left memory region is of lower priority */
+	} else {
+		/* If the left memory region is contained within the right
+		 * region, drop it entirely
+		 */
+		if (RANGE_CONTAIN(mr->pbase, mr->len, ml->pbase, ml->len)) {
+			ml->len = 0;
+
+		/* If the left region has a part of itself in the right region,
+		 * drop that part of the left region only and split by creating
+		 * a new one if the left region is larger than the right region.
+		 */
+		} else {
+			if (RANGE_CONTAIN(ml->pbase, ml->len,
+					  mr->pbase, mr->len))
+				/* Ignore insertion failure as there is nothing
+				 * we can do about it and it is not worth caring
+				 * about.
+				 */
+				ukplat_memregion_list_insert_at_idx(list,
+					&(struct ukplat_memregion_desc){
+						.vbase = mr->pbase + mr->len,
+						.pbase = mr->pbase + mr->len,
+						.len   = ml->pbase + ml->len -
+							 mr->pbase - mr->len,
+						.type  = ml->type,
+						.flags = ml->flags
+					}, ridx + 1);
+
+			ml->len = mr->pbase - ml->pbase;
+		}
+	}
+}
+
+int ukplat_memregion_list_coalesce(struct ukplat_memregion_list *list)
+{
+	struct ukplat_memregion_desc *m, *ml, *mr;
+	int ml_prio, mr_prio;
+	__u32 i;
+
+	i = 0;
+	m = list->mrds;
+	while (i + 1 < list->count) {
+		/* Make sure first that they are ordered. If not, swap them */
+		if (m[i].pbase > m[i + 1].pbase ||
+		    (m[i].pbase == m[i + 1].pbase &&
+		     m[i].pbase + m[i].len > m[i + 1].pbase + m[i + 1].len)) {
+			struct ukplat_memregion_desc tmp;
+
+			tmp = m[i];
+			m[i] = m[i + 1];
+			m[i + 1] = tmp;
+		}
+		ml = &m[i];
+		mr = &m[i + 1];
+		ml_prio = get_mrd_prio(ml);
+		mr_prio = get_mrd_prio(mr);
+
+		/* If they overlap */
+		if (RANGE_OVERLAP(ml->pbase,  ml->len, mr->pbase, mr->len)) {
+			/* If they are not of the same priority */
+			if (ml_prio != mr_prio) {
+				/* If the overlapping memory region of type
+				 * UKPLAT_MEMRT_KERNEL is of lower priority,
+				 * then something is definetely wrong => abort!
+				 */
+				if ((ml_prio < mr_prio &&
```

Comment:
I don't quite follow this, even with the comment, can you please provide an example?